### PR TITLE
CCDB-4880: Adding a S3 Sink Connector config to modify behavior w.r.t tagging errors.

### PR DESCRIPTION
**Pull request by ManasjyotiSharma**
## Problem
[CCDB-4880](https://confluentinc.atlassian.net/browse/CCDB-4880): Databricks Delta Lake Sink Connector sporadically fails with a NPE when trying to read "endOffset" tag from S3 objects. The root cause is that the Delta Lake Sink Connector under the hood leverages S3 Sink Connector code to write/create S3 objects and the S3 Sink Connector code in case of any tagging errors simply ignores them. However from Databricks Sink Connector POV ignoring such tagging errors means the S3 object is left in an inconsistent state because it expects that if object is there in S3 then tag must also be there for that S3 object. See discussions in this [slack thread](https://confluent.slack.com/archives/C011CCECDB7/p1660623278820309).

## Solution
Adding a S3 Sink Connector low level configuartion so that the client/caller can decide whether it is okay for it to ignore tagging errors or not. This PR is part-1 of a 2 parts PR. The part-2 PR will have the Databricks connector side of changes leveraging the newly added S3 Sink config & handle any tagging errors gracefully at Databricks Sink connector end. However until this current S3 Sink PR is merged and a new version is created, I cannot raise part-2 PR as in Databricks Sink Connector pom I'd have to use the new version number of S3 Sink Connector.

##### Does this solution apply anywhere else?
- [x] yes
- [] no

##### If yes, where?
Databricks Delta Lake Sink Connector

## Test Strategy
Manually tested via docker playground the overall tagging failure scenario from Databricks connector end by hardcoding a throw clause in the S3 Sink Connector's TopicPartitionWriter#tagFile method.

##### Testing done:
- [x] Unit tests
- [x] Integration tests
- [ ] System tests
- [x] Manual tests

### CODE REVIEW

**Comment by ManasjyotiSharma** [comment]
See similar/related [PR](https://github.com/confluentinc/kafka-connect-storage-cloud/pull/554). However that was targetting a patch release. Since the changes are not really a bug fix; rather a minor feature addition (new config), so I closed that old PR and creating this new one which will be created as a 10.2.x branch ([reference](https://confluent.slack.com/archives/C011CCECDB7/p1662443508404549?thread_ts=1660623278.820309&cid=C011CCECDB7)) and v10.2.0 will have the changes to be consumed by Delta Lake Sink Connector.

**Comment by ManasjyotiSharma** [comment]
BTW, S3 Sink Connector UTs are not running as part of maven build (reference: [CCMSG-2169](https://confluentinc.atlassian.net/browse/CCMSG-2169)). Seems to be due to some version conflict with surefile and junit-jupiter. I tried fixing the same for sometime by playing around with versions, but at the end could not invest more time on that. I ran the 3 UT Classes that are part of the current PR via Intellij IDEA and all of them passed.

**Comment by ManasjyotiSharma** [comment]
Screenshot from 2022-09-07 10-06-19

File: kafka-connect-s3/src/main/java/io/confluent/connect/s3/S3SinkConnectorConfig.java [hunk 1]
```diff
@@ -1164,7 +1182,7 @@ public String toString() {
     };
 
     public static String[] names() {
-      BehaviorOnNullValues[] behaviors = values();
+      IgnoreFailBehavior[] behaviors = values();
```
Comment (snehashisp) [inline kafka-connect-s3/src/main/java/io/confluent/connect/s3/S3SinkConnectorConfig.java:1185]: nit: 
```suggestion
      IgnoreOrFailBehavior[] behaviors = values();
```

File: kafka-connect-s3/src/main/java/io/confluent/connect/s3/TopicPartitionWriter.java [hunk 2]
```diff
@@ -599,7 +607,11 @@ private void commitFiles() {
       String encodedPartition = entry.getKey();
       commitFile(encodedPartition);
       if (isTaggingEnabled) {
-        tagFile(encodedPartition, entry.getValue());
+        RetryUtil.exponentialBackoffRetry(() -> tagFile(encodedPartition, entry.getValue()),
```
Comment (snehashisp) [inline kafka-connect-s3/src/main/java/io/confluent/connect/s3/TopicPartitionWriter.java:610]: maybe a better place is to add the retry in `storage.addTags(s3ObjectPath, tags)` (line no 658). That we can have retries prior to the ignore of fail check which should help with the robustness

**Comment by snehashisp** [comment]
This looks good. Can you try the pom change I mage for https://github.com/confluentinc/kafka-connect-object-store-source/pull/293/files#diff-9c5fb3d1b7e3b0f54bc5c4182965c4fe1f9023d449017cece3005d3f90e8e4d8 and see if the UTs run (pom.xml changes are at the end of the diff).

File: kafka-connect-s3/src/main/java/io/confluent/connect/s3/S3SinkConnectorConfig.java (hunk 1 above)
Comment (ManasjyotiSharma) [reply kafka-connect-s3/src/main/java/io/confluent/connect/s3/S3SinkConnectorConfig.java]: > nit:

@snehashisp I have incorporated above feedback with [this](https://github.com/confluentinc/kafka-connect-storage-cloud/pull/558/commits/4ac8dccf815e34d4f8540499c55ab60eedeaf3b7) commit.

**Review by pbadani — APPROVED** [review]

**Comment by ManasjyotiSharma** [comment]
> This looks good. Can you try the pom change I mage for https://github.com/confluentinc/kafka-connect-object-store-source/pull/293/files#diff-9c5fb3d1b7e3b0f54bc5c4182965c4fe1f9023d449017cece3005d3f90e8e4d8 and see if the UTs run (pom.xml changes are at the end of the diff).

@snehashisp I have already filed tracking ticket [CCMSG-2169](https://confluentinc.atlassian.net/browse/CCMSG-2169) for that and assigned to @rajat-agarwal. Let current PR-558 be not dependent on that as I have run all tests from IntelliJ and shared the screenshot above.

**Pull request merged**